pkg/registries/registry.go

```diff
@@ -0,0 +1,183 @@
+package registries
+
+import (
+	"bytes"
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+
+	logging "github.com/op/go-logging"
+	"github.com/openshift/ansible-service-broker/pkg/apb"
+	"github.com/openshift/ansible-service-broker/pkg/registries/adapters"
+)
+
+var regex = regexp.MustCompile(`[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*`)
+
+// Config - Configuration for the registry
+type Config struct {
+	URL  string
+	User string
+	Pass string
+	Org  string
+	Type string
+	Name string
+	// Fail will tell the registry that it is ok to fail the bootstrap if
+	// just this registry has failed.
+	Fail      bool     `yaml:"fail_on_error"`
+	WhiteList []string `yaml:"white_list"`
+	BlackList []string `yaml:"black_list"`
+}
+
+// Validate - makes sure the registry config is valid.
+func (c Config) Validate() bool {
+	if c.Name == "" {
+		return false
+	}
+	m := regex.FindString(c.Name)
+	return m == c.Name
+}
+
+// Registry - manages an adapter to retrieve and manage images to specs.
+type Registry struct {
+	config  Config
+	adapter adapters.Adapter
+	log     *logging.Logger
+	filter  Filter
+}
+
+// LoadSpecs - Load the specs for the registry.
+func (r Registry) LoadSpecs() ([]*apb.Spec, int, error) {
+	imageNames, err := r.adapter.GetImageNames()
+	if err != nil {
+		r.log.Errorf("unable to retrieve image names for registry %v - %v",
+			r.config.Name, err)
+		return []*apb.Spec{}, 0, err
+	}
+	// Registry will throw out all images that do not end in -apb
+	imageNames = registryFilterImagesForAPBs(imageNames)
+	validNames, filteredNames := r.filter.Run(imageNames)
+
+	r.log.Debug("Filter applied against registry: %s", r.config.Name)
+
+	if len(validNames) != 0 {
+		r.log.Debugf("Passing APBs:")
+		for _, name := range validNames {
+			r.log.Debugf("-> %s", name)
+		}
+	}
+
+	if len(filteredNames) != 0 {
+		go func() {
+			var buffer bytes.Buffer
+			buffer.WriteString("Filtered APBs:\n")
+			for _, name := range filteredNames {
+				buffer.WriteString(fmt.Sprintf("-> %s", name))
+			}
+			r.log.Infof(buffer.String())
+		}()
+	}
+
+	// Debug output filtered out names.
+	specs, err := r.adapter.FetchSpecs(validNames)
+	if err != nil {
+		r.log.Errorf("unable to fetch specs for registry %v - %v",
+			r.config.Name, err)
+		return []*apb.Spec{}, 0, err
+	}
+	return specs, len(imageNames), nil
+}
+
+func registryFilterImagesForAPBs(imageNames []string) []string {
```

Comment:
@shawn-hurley perfect, thanks @shawn-hurley. We should discuss your thoughts more too, think you're right; there's a better way to handle this.